### ISSUE TITLE
Extremely slow with antivirus

### ISSUE THREAD

**Issue by dizlexik** [eslint/eslint#5679]
The company I work for requires that I run some horrible (but somehow popular) antivirus software on my PC which makes my `eslint .` take nearly 30 seconds to run on a very small project.

I did a little investigating and all of that slowness occurs in `glob-util.js` in the call to `glob.sync(pattern, globOptions)`. If I modify globOptions to include `ignore: 'node_modules/**'` my linting drops to one second.

It appears that crawling node_modules causes my antivirus to kick in for each file, and I'm assuming I can't be the only one with this problem. In any case, filtering out node_modules before the glob.sync call should certainly speed things up to some degree for everyone.

I understand that there is some complex logic involved with ignoring and unignoring paths so I see why all node_modules paths are included at this point in the code, but is there a chance this could be handled in some other way that would short-circuit the needless crawling? It would be a massive 99% speed increase, at least for me and anyone else in my position.

**Comment by eslintbot** [comment]
Hi @dizlexik, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by alberto** [comment]
node_modules is ignored by default. Please provide the information requested

**Comment by dizlexik** [comment]
I understand that node_modules is ignored by default, but `glob.sync(pattern, globOptions)` happens before ignoredPaths rules are applied. My question is whether or not node_modules can be excluded from the glob crawling instead of each file under node_modules being filtered out one-by-one after the fact.

I'm using 2.5.0, but this was also an issue in 2.4.0.

**Comment by ilyavolodin** [comment]
@alberto `node_modules` are ignored by default from linting, but I don't think we ignore them by default for globing. That happens before we even get to ignore options. I think this would be a good enhancement to add, but at the same time, we need to make it in such a way that users would be able to un-ignore `node_modules`, and I don't know if we have enough information to do that at the point of globbing.

**Comment by dizlexik** [comment]
That's exactly right. Thanks for summing it up clearly!

And yes, the un-ignoring is the real complication here. It seems to me that the information on whether or not node_modules or some file(s) under it are to be un-ignored is present at the point of globbing, but I can imagine it might be quite a complicated bit of potentially error-prone code to jam in there to fix this :/

Hopefully someone can find the time to tackle this because I currently have a build process that runs across a number of sub-modules that each perform linting individually and this would reduce the build time from minutes to seconds. I've patched eslint on our private npm server with this hack for now since we don't do any un-ignoring, but an official fix would be awesome.

**Comment by alberto** [comment]
Sorry, I didn't read close enough.

**Comment by nzakas** [comment]
@dizlexik since you've already done a bit of research here, would you be willing to submit a pull request? We are pretty strapped with some high priority issues right now, so it's unlikely that we will find the time to dig into this anytime soon.

**Comment by dizlexik** [comment]
I'd love to help out, but after reviewing the code in this area I don't see any easy way of determining whether or not any node_modules paths are unignored, so it looks to me like some changes will be required at a lower level.

I see that `ingored-files.js utilizes the `ignore` package, which doesn't appear to expose anything useful regarding this so some manual testing of each ignore pattern may be required to resolve this. Sounds like a mini-project that I don't have the time for right now, I'm sorry :/

**Comment by nzakas** [comment]
Unfortunately, it's unlikely to get done unless we get a volunteer. I'll flag it as such to see if we get any takers.

**Comment by maxtruxa** [comment]
From the looks of it this could be a huge performance improvement. Provided nobody else wants to solve it in the near future, I'll take a look at it until the end of the week.

**Comment by alberto** [comment]
Note `.eslintignore` uses `ignore` pattern syntax, while `glob` uses `minimatch`.

**Comment by holm** [comment]
Could `node_modules` just not always be ignored? It makes a huge difference to the speed of eslint, and I have a hard time seeing the use-case for linting `node_modules`.

**Comment by nzakas** [comment]
@holm sorry, I can't quite parse your meaning ("just not always"). We ignore `node_modules` by default already.

Ultimately, we need some real data here to move forward. By data, I mean we need a benchmark that shows what we're doing now is slower than something we could potentially be doing.

@maxtruxa are you still planning on investigating this?

**Comment by alberto** [comment]
@nzakas I think he means in glob walking

**Comment by holm** [comment]
Sorry for the unclear comment. Yes, I mean in the glob walking. In large projects, with many dependencies, `node_modules` can contain _a lot_ of files. One project I am currently working on has 43,432 files in `node_modules`.

**Comment by alberto** [comment]
@holm [here](https://github.com/eslint/eslint/issues/5410) is an example of someone needing to lint a folder inside node_modules

**Comment by holm** [comment]
Quite surprising. I would still argue that having your a modules source code in node_modules is a bad idea, I can appreciate that it's better to be a little slower and flexible, than fast and inflexible.

For my current case excluding node_modules from glob, moves the eslint time from about 7500 ms to 700 ms, which makes a huge difference when running tests and lints very often.

In any case thanks a lot for eslint, it is an amazing tool!

**Comment by ilyavolodin** [comment]
@holm We just implemented a change in our .eslintignore file to allow un-ignoring things like node_modules. It would be counter-productive to now remove ability for globs to pick up files in those directories:-) The issue is accepted, so we recognize and agree that this is a problem, but nobody has gotten around to fixing it yet.
I think the right way to fix this would be to have .eslintignore file also drive exclusion patterns for globing, so that whatever you specify there (plus our defaults) would be ignored. I'm not 100% sure how easy (or even possible) it would be to implement, since there might be timing issue there.

**Comment by IanVS** [comment]
One possibility here would be to continue ignoring node_modules by default (as we currently do), but to override that behavior instead of using a negating ignore pattern (like `!node_modules/*`), we could add a cli flag (`--no-default-ignores`) to disable the default ignores entirely, leaving it up to the user to specify the ignores they want.  Then, in the absence of this flag, we could avoid crawling node_modules and bower_components.

I've been spending a fair bit of time in our ignore logic lately, and without something like what I'm proposing I think it will be difficult to correctly accomplish what's being discussed in this issue (due to the problems raised by @alberto and @ilyavolodin).

**Comment by alberto** [comment]
Does anyone know what are the differences in behaviour between gitignore and glob? maybe you @kaelzhang ?

**Comment by nzakas** [comment]
@IanVS We could also do some fuzzy logic and avoid crawling `node_modules` and `bower_components` if either:

1. There is no ignore file and no `--ignore-pattern`
2. The ignore file or ignore pattern do not contain `node_modules` or `bower_components`

I'd really rather not add any more ignore-related flags if we can avoid it.

**Comment by IanVS** [comment]
@nzakas, sure that would at least cover most cases.  👍  from me.

**Comment by kaelzhang** [comment]
@alberto gitignore pattern acts as the glob pattern only if there are no slashes(`/`), such as `foo*`, or `abc/*.js`. gitignore specifies a certain behavior of two asterisks partially, because the things which two asterisks match varies cross platforms.

I'd like to be the volunteer. I could do it just during the vacation of Labours' Day.

The optimization must be done at the very stage of traversing directories. If a directory is ignored, it should be skipped to traverse down into that directory.

Let me spelunk into the code of glob first.

**Comment by alberto** [comment]
@kaelzhang Thanks. Do you intend to implement what @nzakas proposed or something different? If the latter, can you explain it so we can choose the best alternative?

**Comment by kaelzhang** [comment]
The performance issue is encountered because we use glob to get a full list of files, then filter them with the gitignore rules.

To solve the problem, we do not need to add more flags, just to optimize the glob-and-filter process.

### If we ignore node_modules
(either by default or my custom ignore config)

1. traverse the project root to get lib, index.js, and node_modules, for example
- index.js is a FILE, end
- lib is a DIR, deep down into it
- node_modules is a DIR, but it is ignored, so skip it and avoid decending down even further

2. traverse lib

...

### If something inside node_modules should be re-

**Comment by ilyavolodin** [comment]
@kaelzhang To the best of my recollection, we don't traverse directories ourselves, `node-glob` is doing it for us, so the way you are suggesting we fix it would either require us to stop using `node-glob`, or modify it somehow. I think glob allows you to pass directories to be ignored before you call `resolve`, so that's probably the easiest way to deal with this issue.

**Comment by mdevils** [comment]
Working on this (hope I can handle it).

**Comment by kaelzhang** [comment]
It is not that easy.

1. The inquirer prompt of `patterns` supports to pass a glob pattern to tell `eslint` what code to examine, [here](https://github.com/eslint/eslint/blob/master/lib/config/config-initializer.js#L143). The patterns user passed to eslint also support minimatch-customized glob patterns, such as `a{1..5}b`(bracing), `a?(b*)`(state characters), etc.
2. `node-glob` does not provide a way to specify custom filters to filter out unnecessary directories/files during the traversing process.

It is blocked by {2}, or it needs to create another `glob`-like package to do this, which will be a big project.

**Comment by Zarel** [comment]
I was told to comment about my problem, so:

I have a `logs/` directory with around 100 million files in it. It's excluded using `.eslintignore`, but eslint's behavior of loading a list of every file in it before applying ignore patterns is making it run out of memory.

Example crash:

```
$ eslint --rulesdir=dev-tools/eslint --debug --cache .
>>
<< undefined
<--- Last few GCs --->
82716 ms: Scavenge 1401.6 (1443.1) -> 1401.6 (1443.1) MB, 3.2 / 0 ms (+ 423.2 ms in 1 steps since last GC) [allocation failure] [incremental marking delaying mark-sweep].
84038 ms: Mark-sweep 1401.6 (1443.1) -> 1400.7 (1443.1) MB, 1321.7 / 0 ms (+ 909.8 ms in 18 steps since start of marking, biggest step 423.2 ms) [last resort gc].
85355 ms: Mark-sweep 1400.7 (1443.1) -> 1400.7 (1443.1) MB, 1316.9 / 0 ms [last resort gc].
<--- JS stacktrace --->
==== JS stack trace =========================================
Security context: 0x35e6b44d81b1 <JS Object>
2: _readdir [/home/ps/showdown/node_modules/glob/sync.js:275] [pc=0x377bb7434d3d] (this=0x34a8bbb2e259 <a GlobSync with map 0x31dff91a84f9>,abs=0xfb241d852a9 <String[44]: /home/ps/showdown/logs/2015-11/ou/2015-11-14>,inGlobStar=0x35e6b4404299 <false>)
3: _readdirInGlobStar [/home/ps/showdown/node_modules/glob/sync.js:254] [pc=0x377bb7437acf] (this=0x34a8bbb2e259 <a GlobSync with map 0x3...
Tue, 03 May 2016 02:13:25 GMT eslint:cli Running on files
Tue, 03 May 2016 02:13:25 GMT eslint:cli-engine Loading rules from dev-tools/eslint
Tue, 03 May 2016 02:13:25 GMT eslint:ignored-paths Looking for ignore file in /home/ps/showdown
Tue, 03 May 2016 02:13:25 GMT eslint:ignored-paths Loaded ignore file /home/ps/showdown/.eslintignore
Tue, 03 May 2016 02:13:25 GMT eslint:ignored-paths Adding /home/ps/showdown/.eslintignore
Tue, 03 May 2016 02:13:25 GMT eslint:glob-util Creating list of files to process.
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
```

**Comment by nzakas** [comment]
@kaelzhang node-glob does have an `ignore` option that lets you specify directories it should not traverse. It's buried in the list of options here: https://github.com/isaacs/node-glob#options

**Comment by kaelzhang** [comment]
@nzakas Yes, I've read the source code of `glob` and `minimatch`. 

But `options.ignore` will be mapped into an array of `minimatch` instances. I thought it it not that logical to use gitignore patterns as `minimatch` patterns.

**Comment by nzakas** [comment]
@kaelzhang you don't have to. My suggestion was only to do that in the couple cases that we know we don't have to search the default directories, so just use "Node_modules" and "bower_components" for `options.ignore`. There's no need to do anything more complicated.

**Comment by ilyavolodin** [comment]
@nzakas I'm not sure I agree. We just merged a change that allows users to un-ignore directories ignored by default. If we were to hard-code `node_modules` and `bower_components` as ignored by `node_glob`, that would invalidate previous change.

**Comment by platinumazure** [comment]
@ilyavolodin One of these modules allows us to work with strings only without going into the filesystem, right? If so, is it possible to run `node_modules/`, `bower_components/`, and the user's ignore patterns through that logic to see if anything in `node_modules` or `bower_components` needs to be traversed at all? And if so, we can specify skip-traverse for those directories?

**Comment by Zarel** [comment]
One first step would be to hard-code all the ignored directories if there are no unignored directories.

**Comment by nzakas** [comment]
@ilyavolodin please read my [previous comment](https://github.com/eslint/eslint/issues/5679#issuecomment-215831547). What I'm suggesting would not interfere with re-including.

**Comment by alberto** [comment]
I'm working on a simple fix for this for the default ignores, as mentioned [here](https://github.com/eslint/eslint/issues/5679#issuecomment-215831547)

**Comment by Zarel** [comment]
This doesn't solve my own problem, which is that the huge ignored directory is `logs/`, not `node_modules/`...

**Comment by ilyavolodin** [comment]
@Zarel We are aware that the fixed that we just merged doesn't solve all of the problems. It's a temporary solution that should help most of the users. We will be working on the more permanent solution.

**Comment by Zarel** [comment]
It's just that this issue was closed. Should it be opened, or is there a different issue that tracks that problem?

**Comment by platinumazure** [comment]
@Zarel Very sorry for the delayed response. We should do further work under a new issue, perhaps titled something like "Skip glob traversal of ignored directories (besides node_modules and bower_components)". Thanks!

**Comment by Zarel** [comment]
Sure. Are you planning on opening the issue or should I?

**Comment by rocketmonkeys** [comment]
@platinumazure @ilyavolodin @Zarel - I've made a follow-up issue to improve on this fix here: #6710 .  I also have a very rough PR (that I'm sure isn't good yet) that I can link to on that issue.
